Use different method to hide menus

brezocordero opened the pull request:
## Description

`add_submenu_page` with empty `parent_slug` generates a deprecation error.
This PR uses a non-existing parent slug instead of an empty string.

## Code review notes

We could also wait for the trac ticket but not sure if it will be implemented or when.

## QA notes

2 ways to test it:
- On Jurassic.ninja, click on 'Select More Options', select 'PHP 8.1', WooCommerce, and create a new site.
- Install plugin Query Monitor.
- Upload the build
- Check there are no errors in Query Monitor when you go to MailPoet landing page

And locally:
- On your local site, if running PHP 8.1, you can test there is no log by going to:
- ?page=mailpoet-welcome-wizard
- ?page=mailpoet-landingpage
- ?page=mailpoet-woocommerce-setup

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5718]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5716]: https://mailpoet.atlassian.net/browse/MAILPOET-5716?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MAILPOET-5718]: https://mailpoet.atlassian.net/browse/MAILPOET-5718?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes commented on mailpoet/lib/Config/Menu.php:
```diff
@@ -213,8 +221,58 @@ private function registerMailPoetMenu() {
 
-    $this->registerMailPoetSubMenuEntries();
+  /**
+   * Hide the submenu page based on slug and return the item that was hidden.
```
The method doesn't seem to be returning anything, even though it is mentioned here (and again a few lines below).

brezocordero replied:
Modified

JanJakes commented on mailpoet/lib/Config/Menu.php:
```diff
@@ -213,8 +221,58 @@ private function registerMailPoetMenu() {
+   * Borrowed from Jetpack's Base_Admin_Menu class.
+   *
+   * Instead of actually removing the submenu item, a safer approach is to hide it and filter it in the API response.
+   * In this manner we'll avoid breaking third-party plugins depending on items that no longer exist.
```
From this description, it seems that this is a workaround for calling `remove_submenu_page`. Since we're not hiding any WP or 3rd party pages, but only our own, and we're not doing that conditionally, but always, could we simply call `remove_submenu_page` at a correct moment?

brezocordero replied:
With `remove_submenu_page` you can not access the page but we still want the pages to be available.

JanJakes commented on mailpoet/lib/Config/Menu.php:
```diff
@@ -213,8 +221,58 @@ private function registerMailPoetMenu() {
+   * @param string $submenu_slug The submenu slug that should be hidden.
+   * @return void
+   */
+  public function hideSubmenuPage(string $menu_slug, string $submenu_slug): void {
```
The naming of the variables (`$menu_slug`, etc.) doesn't follow our coding standards, but, interestingly, the lint passes? 🤔

brezocordero replied:
Thanks for noticing. Changed.

JanJakes commented on mailpoet/lib/Config/Menu.php:
```diff
@@ -61,6 +61,7 @@ class Menu {
   const LANDINGPAGE_PAGE_SLUG = 'mailpoet-landingpage';
 
   const ICON_BASE64_SVG = 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxNTIuMDIgMTU2LjQiPjxwYXRoIGZpbGw9IiNhN2FhYWQiIGQ9Ik0zNy43MSw4OS4xYzMuNSwwLDUuOS0uOCw3LjItMi4zYTgsOCwwLDAsMCwyLTUuNFYzNS43bDE3LDQ1LjFhMTIuNjgsMTIuNjgsMCwwLDAsMy43LDUuNGMxLjYsMS4zLDQsMiw3LjIsMmExMi41NCwxMi41NCwwLDAsMCw1LjktMS40LDguNDEsOC40MSwwLDAsMCwzLjktNWwxOC4xLTUwVjgxYTguNT ...
+  const HIDE_CSS_CLASS = 'hide-if-js';
```
What does the class name mean? Is it a native class from WP?

brezocordero replied:
Yes,[ it is in core](https://github.com/WordPress/wordpress-develop/blob/30954f7ac0840cfdad464928021d7f380940c347/src/wp-admin/css/common.css#L176). It is used by [WPCOM API to hide the menus from the JS navigation](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L237).
It is also [the approach that is used in Woo Express to hide the menus](https://github.com/Automattic/wc-calypso-bridge/blob/20af7c76771c7bbe70b825e17a25a46c2d5d3111/includes/class-wc-calypso-bridge-ecommerce-admin-menu.php).

JanJakes reviewed (commented):
It took me some time to realize what is the issue here, but I think I get it now. We're talking only about pages, that should have no parent, and that's only a few pages now, is that correct?

What if we added them under a menu item, e.g. the MailPoet homepage? It would highlight the homepage in the menu, though, not sure if that's desired.

Alternatively, what if we use `unknown-parent-slug` or something like that? Would that fail with a warning as well?

I'm adding other remarks and questions in the code.

brezocordero commented:
@JanJakes I have fixed the variable names and the return of the function
I have explained a bit more about this approach in the comments.
To sum up: 
We hide the menus with CSS using a specific class `.hide-if-js` that is present in WP core. 
The class is also used to filter which menus are displayed on the WPCOM API/JS navigation.
The pages can still be accessed using the direct link.

JanJakes committed:
Fix deprecation error for pages hidden in menu

[MAILPOET-5718]

JanJakes commented:
@brezocordero Thanks, for the context, I think I understand what's happening now. The issue seems to occur only when you access one of the hidden pages, not when the menu is rendered. And it seems to me that the empty string really is the only problem, so I wonder, why not keeping the current method, and using just a simple fix such as [this](https://github.com/mailpoet/mailpoet/commit/2a938fcf4c22efdd1b9563a4d626336fa86a945a)?

What confuses me in the PR is that the current hiding method is still somehow kept — the `unset($submenu[self::MAIN_PAGE_SLUG]);` and `$plugin_page = self::MAIN_PAGE_SLUG;` (actually even added to one more place) — while we also add hiding with CSS. The code then becomes pretty hard to understand (and it seems we don't need the CSS workaround at all).

As I was a little bit struggling with understanding and replicating the issue, I'd also suggest extending the QA notes with something like this:

> Test this under PHP 8.1 or higher (the error won't be visible on Jurassic Ninja, which runs PHP 8.0).
> Check that this PR fixes the error on the following pages:
>   1. `?page=mailpoet-welcome-wizard`
>   2. `?page=mailpoet-landingpage`
>   3. `?page=mailpoet-woocommerce-setup`
>   4. MailPoet error page — there's no easy way to test this, so no need.

brezocordero commented:
> why not keeping the current method, and using just a simple fix such as [this](https://github.com/mailpoet/mailpoet/commit/2a938fcf4c22efdd1b9563a4d626336fa86a945a)?

Ok, makes sense. Thanks a lot, I also like it better like this.
I have taken your commit, added a bit that was missing and updated the test.
Updated the PR description and the QA instructions.

JanJakes reviewed (approved):
@brezocordero Thanks! It's good to know about the CSS solution, in case we'd run into any issues with our current solution again.

Aschepikov reviewed (approved):
GJ!
